README
NAME
    Tangerine - Analyse perl files and report module-related information

SYNOPSIS
        use Tangerine;
        use version 0.77;

        my $scanner = Tangerine->new(file => $file, mode => 'all');
        $scanner->run;

        print "$file contains the following modules: ".
            join q/, /, sort keys %{$scanner->provides}."\n";

        print "$file requires Exporter on the following lines: ".
            join q/, /, sort map $_->line, @{$scanner->requires->{Exporter}}."\n";

        my $v = 0;
        for ( @{$scanner->uses->{Test::More}}) {
            $v = $_->version if $_->version && qv($v) < qv($_->version)
        }
        print "The minimum version of Test::More required by $file is $v\n";

DESCRIPTION
    Tangerine statically analyses perl files and reports various information
    about provided, used (compile-time dependencies) and required (runtime
    dependencies) modules.

    Currently, PPI is used for the initial parsing and statement extraction.

CONSTRUCTOR
    "new"
        Creates the Tangerine object. Takes the following two named
        arguments:

            'file', the file to analyse
            'mode', what should we look for; may be one of 'all', 'prov', 'dep',
                'req' or 'use'.  'dep' implies both 'req' and 'use'.  Single
                letter abbreviations are also accepted.

        Both arguments are optional, however, 'file' needs to be set before
        running the scanner, e.g.

            my $scanner = Tangerine->new;
            $scanner->file($file);
            $scanner->run;

METHODS
    "run"
        Runs the analysis.

    "provides"
        Returns a hash reference. Keys are the modules provided, values
        references to lists of Tangerine::Occurence objects.

    "requires"
        Returns a hash reference. Keys are the modules required at run-time,
        values references to lists of Tangerine::Occurence objects.

    "uses"
        Returns a hash reference. Keys are the modules required at
        compile-time, values references to lists of Tangerine::Occurence
        objects.

SEE ALSO
    Tangerine::Occurence

AUTHOR
    Petr Šabata <contyk@redhat.com>

COPYRIGHT AND LICENSE
    Copyright (c) 2014-2015 Petr Šabata

    See LICENSE for licensing details.



Changes
Revision history for Perl module Tangerine

0.22 2016-02-25
- Handle the version module regexes better and define our own simple fallback
- Recommend List::Util 1.33+ but just use List::MoreUtils if it isn't available
- Added support for the simple mixin class
- Removed a useless debug print

0.21 2016-01-25
- Restore compatibility with perl 5.10.

0.20 2016-01-22
- Don't crash if the hook isn't really loaded
- Switched to using List::Util
- Ignore more false positives (e.g. use base __PACKAGE__->...)
- Added support for Test::Class::Most
- Added support for Role::Tiny::With
- Various minor speed optimisation all over the place

0.19 2015-08-20
- HookData now allows to override the hook's own type.
- Documentation for HookData was corrected and updated.
- The anymoose hook was dropped; it didn't really make much sense.
- This release also introduces a new hook, adding support for Module::Hook.
- The inline hook was enhanced to properly detect Inline::Filters.
- And finally, the list hooks now support Mojo::Base, mixin::with and Tk::widgets.

0.18 2015-06-24
- The Inline hook now supports language aliases.

0.17 2015-05-15
- Added [GithubMeta] to dist.ini, so the repo will be listed in the
  dist metadata, and also [MetaJSON], so the release will include a
  META.json file.  Thanks to Neil Bowers
- Ported the testsuite to the new API introduced in 0.15

0.16 2015-05-13

- No longer ignore numeric module names.  Apparently it's allowed in
  some situations.

0.15 2015-04-24

- The utility now lives in its own distribution, App::Tangerine
- Renamed provides/requires/uses methods to package/require/compile;
  the old method names are still available but consider them deprecated
- Renamed the hook types respectively, from prov/req/use to
  package/require/compile; this is an internal change
- Mode accepts the new type names; also, these are now preferred
- Drop Mo in favour of a simple object system, cutting the dependency chain
- Simplified module name filtering so it actually works like it should
- Fixed a nasty bug with explicit zero version requirements
- Updated documentation

0.14 2015-03-30

 - Mooselike should now parse extends/with statements enclosed in parens
   somewhat better

0.13 2015-02-24

 - Inline is now processed but its own specialized hook, hopefully supporting
   all the import cases properly (for our purposes at least)
 - Added support for Module::Runtime and its subs
 - POD fixes

0.12 2015-01-12

 - Add support for Test::Requires and superclass
 - Update copyright notices
 - stripquotelike() now explodes anonymous hashes too
 - The XXX hook now doesn't require YAML on empty imports
 - `use Inline ()' now works somewhat better

0.11 2014-11-26

 - Moose-like support rewritten -- we now detect versions and avoid indirect
   object method calls
 - aliased fixed; it now only reports the first module in the list, duh
 - We no longer report obvious filenames, really
 - The required version of perl can be an integer as well, duh again
 - List constructors spread over multiple lines are now processed correctly
 - Package names ending in :: are now reported without this suffix

0.10 2014-10-15

 - New internal hook API
 - Support for use_ok, require_ok, syntax_ok from various test distributions,
   and for "use ok" added
 - Inline support added
 - XXX support added
 - Minor performance improvements
 - We no longer report reqs/uses containing non-module strings, like
   filesnames or variables

0.06 2014-10-06

 - Restore compatibility with 5.10

0.05 2014-09-29

 - Release with proper POD

0.04 2014-09-29

 - Detect conditional and compound require statement correctly
 - Support Any::Moose
 - Support Moose-like inheritence (`with' and `extends')

0.03 2014-09-08

 - Add ExecDir to dist.ini to install bin/tangerine by default

0.02 2014-09-05

 - Switch to PP Mo
 - Include the script's deps in prereqs

0.01 2014-09-04

 - This is the initial public release


lib/Tangerine.pm
package Tangerine;

# ABSTRACT: Examine perl files and report dependency metadata
use 5.010;
use strict;
use warnings;
use utf8;
use PPI;
use Scalar::Util qw(blessed);
use Tangerine::Hook;
use Tangerine::Occurence;
use Tangerine::Utils qw(any accessor addoccurence none);

sub new {
    my $class = shift;
    my %args = @_;
    bless {
        _file => $args{file},
        _mode => $args{mode} // 'all',
        _hooks => {
            package => [ qw/package/ ],
            compile => [ qw/use list prefixedlist if inline moduleload
                moduleruntime mooselike testrequires tests xxx/ ],
            runtime => [ qw/require/ ],
        },
        _package => {},
        _compile => {},
        _runtime => {},
    }, $class
}

sub file { accessor _file => @_ }
sub mode { accessor _mode => @_ }

sub package { accessor _package => @_ }
sub compile { accessor _compile => @_ }
sub runtime { accessor _runtime => @_ }
# For pre-0.15 compatibility
*provides = \&package;
*requires = \&runtime;
*uses = \&compile;

sub run {
    my $self = shift;
    return 0 unless -r $self->file;
    $self->mode('all')
        unless $self->mode =~
            /^(a(ll)?|p(ackage|rov)?|compile|d(ep)?|r(untime|eq)?|u(se)?)$/;
    my $document = PPI::Document->new($self->file, readonly => 1);
    return 0 unless $document;
    my $statements = $document->find('Statement') or return 1;
    my @hooks;
    for my $type (qw(package compile runtime)) {
        for my $hname (@{$self->{_hooks}->{$type}}) {
            my $hook = "Tangerine::hook::$hname";
            if (eval "require $hook; 1") {
                push @hooks, $hook->new(type => $type);
            } else {
                warn "Couldn't load the tangerine hook `${hname}'!";
            }
        }
    }
    @hooks = grep {
            if ($self->mode =~ /^a/o ||
                $_->type eq 'package' && $self->mode =~ /^p/o ||
                $_->type eq 'compile' && $self->mode =~ /^[cdu]/o ||
                $_->type eq 'runtime' && $self->mode =~ /^[dr]/o) {
                $_
            }
        } @hooks;
    my $children;
    my $forcetype;
    STATEMENT: for my $statement (@$statements) {
        $children //= [ $statement->schildren ];
        if ($children->[1] &&
            ($children->[1] eq ',' || $children->[1] eq '=>')) {
            undef $children;
            next STATEMENT
        }
        for my $hook (@hooks) {
            if (my $data = $hook->run($children)) {
                my $modules = $data->modules;
                undef %$modules if any { $_ eq '->' } keys %$modules;
                for my $k (keys %$modules) {
                    if ($k !~ m/^[a-z_][a-z0-9_]*(?:::[a-z0-9_]+)*(?:::)?$/io ||
                        $k =~ m/^__[A-Z]+__$/o) {
                        delete $modules->{$k};
                        next
                    }
                    if (my ($class) = ($k =~ /^(.+)::$/o)) {
                        $modules->{$class} = $modules->{$k}
                            unless exists $modules->{$class};
                        delete $modules->{$k};
                        $k = $class
                    }
                    $modules->{$k}->line($statement->line_number);
                }
                my $type = $forcetype // $hook->type;
                if ($type eq 'package') {
                    $self->package(addoccurence($self->package, $modules));
                } elsif ($type eq 'compile') {
                    $self->compile(addoccurence($self->compile, $modules));
                } elsif ($type eq 'runtime') {
                    $self->runtime(addoccurence($self->runtime, $modules));
                }
                if (@{$data->hooks}) {
                    for my $newhook (@{$data->hooks}) {
                        next if ($newhook->type eq 'package') && ($self->mode =~ /^[dcru]/o);
                        next if ($newhook->type eq 'runtime') && ($self->mode =~ /^[pcu]/o);
                        next if ($newhook->type eq 'compile') && ($self->mode =~ /^[pr]/o);
                        push @hooks, $newhook
                            if none {
                                blessed($newhook) eq blessed($_) &&
                                $newhook->type eq $_->type
                            } @hooks;
                    }
                }
                if (@{$data->children}) {
                    $children = $data->children;
                    $forcetype = $data->type;
                    redo STATEMENT;
                }
            }
        }
        undef $children,
        undef $forcetype;
    }
    1;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Tangerine - Examine perl files and report dependency metadata

=head1 SYNOPSIS

    use Tangerine;
    use version 0.77;

    my $scanner = Tangerine->new(file => $file, mode => 'all');
    $scanner->run;

    print "$file contains the following modules: ".
        join q/, /, sort keys %{$scanner->package}."\n";

    print "$file requires Exporter, at runtime, on the following lines: ".
        join q/, /, sort map $_->line, @{$scanner->runtime->{Exporter}}."\n";

    my $v = 0;
    for (@{$scanner->compile->{'Test::More'}}) {
        $v = $_->version if $_->version && qv($v) < qv($_->version)
    }
    print "The minimum version of Test::More required by $file is $v.\n";

=head1 DESCRIPTION

Tangerine examines perl files and reports dependency metadata -- provided
modules, and both compile-time and run-time dependencies, along with line
numbers, versions and possibly other related information.

Currently, PPI is used for the initial parsing and statement extraction.

=head1 CONSTRUCTOR

=over

=item C<new>

Creates the Tangerine object.  Takes the following two named arguments:

    'file', the file to examine
    'mode', determines what to look for; may be one of 'all',
        'package', 'compile', or 'runtime'.

Both arguments are optional, however, 'file' needs to be set before
running the scanner, e.g.

    my $scanner = Tangerine->new;
    $scanner->file($file);
    $scanner->run;

=back

=head1 METHODS

=over

=item C<run>

Runs the analysis.

=item C<package>

Returns a hash reference.  Keys are the modules provided, values
references to lists of L<Tangerine::Occurence> objects.

=item C<compile>

Returns a hash reference.  Keys are the modules required at compile-time,
values references to lists of L<Tangerine::Occurence> objects.

=item C<runtime>

Returns a hash reference.  Keys are the modules required at run-time,
values references to lists of L<Tangerine::Occurence> objects.

=item C<provides>

=item C<requires>

=item C<uses>

Deprecated.  These are provided for backwards compatibility only.

=back

=head1 SEE ALSO

L<Tangerine::Occurence>

=head1 REPOSITORY

L<https://github.com/contyk/tangerine.pm>

=head1 AUTHOR

Petr Šabata <contyk@redhat.com>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2014-2016 Petr Šabata

See LICENSE for licensing details.

=cut
